### PR TITLE
Add API Handlers to existing implementation

### PR DESCRIPTION
**Pull request by GAlexIHU**
## Overview

OAPI
- Aliases were changed to unions for better name generation
- Input types were changed as they were missing some fields

Add API handlers for
- create (based on plan, no custom subscription support yet)
- view
- edit running (via patches)

Changes startup
- So loading plan & subscription handlers is decided directly by config value

Subscriptions
- Collect and return all errors from validations (only a single error is returned through API as is, but statuses are mapped, error handling will be a separate discrete effort as its a bit complex for Subscriptions)
- Added missing Name, Description & Metadata fields. In some scenarios it's not clear where they should be sourced from, that can be resolved later.

## Notes for reviewer

- I intend to revisit patch serialization and parsing later, as currently its a pain
- PlanAdapter will be reworked once ProductCatalog models are available

- [ ] E2E API tests

### CODE REVIEW

**Comment by GAlexIHU** [comment]
@solkimicreb FYI I had to change parts of TypeSpec, could be relevant for your test integration

File: e2e/productcatalog_test.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+		apiRes, err := client.EditSubscriptionWithResponse(ctx, subscriptionId, api.EditSubscriptionJSONRequestBody{
+			Customizations: []api.SubscriptionEditOperation{o1, o2},
+		})
+		require.Nil(t, err)
+
+		assert.Equal(t, 200, apiRes.StatusCode(), "received the following body: %s", apiRes.Body)
+	})
+
+	t.Run("DUMMY", func(t *testing.T) {
```
Comment (tothandras) [inline e2e/productcatalog_test.go:297]: Do we need this? :)

File: api/spec/src/productcatalog/subscription.tsp [hunk 2]
```diff
@@ -229,109 +269,86 @@ model PlanSubscriptionCreate {
...
+enum RemovePhaseShifting {
+  /**
+   * Shifts all subsequent phases to start sooner by the deleted phase's length
+   */
+  Next,
+
+  /**
+   * Extends the previous phase to end later by the deleted phase's length
+   */
+  Prev,
```
Comment (tothandras) [inline api/spec/src/productcatalog/subscription.tsp:353]: Provide values to the enum values.

```suggestion
  Next: "next",

  /**
   * Extends the previous phase to end later by the deleted phase's length
   */
  Prev: "previous",
```

File: api/spec/src/productcatalog/routes.tsp [hunk 3]
```diff
@@ -267,9 +251,10 @@ interface Subscriptions {
   @operationId("createSubscription")
   @summary("Create subscription")
   @extension("x-internal", true)
-  create(
-    @body body: SubscriptionCreate,
-  ): Subscription | OpenMeter.ConflictError | OpenMeter.CommonErrors;
+  create(@body body: SubscriptionCreate): {
+    @statusCode _: 201;
+    @body body: Subscription;
```
Comment (tothandras) [inline api/spec/src/productcatalog/routes.tsp:256]: Would it make sense to make `name` optional or omitted in `PlanSubscriptionChange` and use the plan's name by default? Also, currency should come from plan and could be removed.

File: openmeter/subscription/httpdriver/edit.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+				return EditSubscriptionRequest{}, err
+			}
+
+			var body api.EditSubscriptionJSONRequestBody
+
+			if err := commonhttp.JSONRequestBodyDecoder(r, &body); err != nil {
+				return EditSubscriptionRequest{}, err
+			}
+
+			if body.Customizations == nil {
```
Comment (tothandras) [inline openmeter/subscription/httpdriver/edit.go:41]: Maybe?
```suggestion
			if len(body.Customizations) == 0 {
```

**Review by tothandras — APPROVED** [review]

**Review by tothandras — APPROVED** [review]

**Review by tothandras — APPROVED** [review]

**Review by tothandras — APPROVED** [review]

**Pull request merged**